Refactor/Improve InvokeAsync to address potential disposing issues

KlausLoeffelmann opened the pull request:
Addresses #12696, #12697 
(Mentioning the latter just for same topic. I do not think, this is an issue really.)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/winforms/pull/13564)

KlausLoeffelmann committed:
Refactor/Improve InvokeAsync to address disposing issues of the cancellation token.

KlausLoeffelmann committed:
Add unit tests for Control.InvokeAsync.

kirsan31 commented on src/System.Windows.Forms/System/Windows/Forms/Control_InvokeAsync.cs:
```diff
@@ -162,31 +163,37 @@ public async Task InvokeAsync(Func<CancellationToken, ValueTask> callback, Cance
-            BeginInvoke(async () => await WrappedCallbackAsync().ConfigureAwait(false));
-            await completion.Task.ConfigureAwait(false);
-        }
+        BeginInvoke(async () => await WrappedCallbackAsync());
```
I agree with #12697 that it would be better to use `BeginInvoke(WrappedCallbackAsync);` here.

kirsan31 commented on src/System.Windows.Forms/System/Windows/Forms/Control_InvokeAsync.cs:
```diff
@@ -226,25 +233,37 @@ public async Task<T> InvokeAsync<T>(Func<CancellationToken, ValueTask<T>> callba
-            BeginInvoke(async () => await WrappedCallbackAsync().ConfigureAwait(false));
-            return await completion.Task.ConfigureAwait(false);
-        }
+        BeginInvoke(async () => await WrappedCallbackAsync());
```
Same as above: `BeginInvoke(WrappedCallbackAsync);`

KlausLoeffelmann committed:
Update docs.

KlausLoeffelmann committed:
Finalize unit tests.

Improve invoke behavior in InvokeAsync.

Update folder for Copilot instructions for generating the InvokeAsync unit tests.

(Note: They had been manually refactored quite a lot, but still saved tons of time.)

KlausLoeffelmann committed:
Add review suggestions and improve docs accordingly.

Shyam-Gupta commented on src/System.Windows.Forms/System/Windows/Forms/Control_InvokeAsync.cs:
```diff
@@ -73,38 +109,66 @@ void WrappedAction()
+    ///    <description>
+    ///     Always pass a <see cref="CancellationToken"/> that you cancel when the
+    ///     control is disposing/its handle is destroyed (recommended). A common pattern is to
+    ///     use a token linked to the control's lifetime.
```
`A common pattern is to use a token linked to the control's lifetime.`
Which pattern are we referring to here?

KlausLoeffelmann replied:
Funny, you would mention that.
I discussed that in depth:

**Pro:** We could have an internal factory, which links the cancellation token to the control, and then captures the `Disposed` or the `HandleDestroyed` event, and _then_ cancels automatically, when that happens.

**Con:** To blackbox-ish. And might be counterproductive, because the token could be then potentially preventing the whole control from being collected, because of the event wiring.

That why I added this to the docs, and we are taking the cancellation token from the outside, or do not use one.

KlausLoeffelmann replied:
Oh, and which pattern is that?
A factory class for the token which does exactly that, so, create a token, wire it, so it cancels the task on HandleDestroyed/Dispose, and then provide the token for other cancel options.

Would you mention the approach there?

KlausLoeffelmann replied:
Agree. It's either the code in the docs, or dropping it.
And the code would be too much here.

Shyam-Gupta commented on src/System.Windows.Forms/System/Windows/Forms/Control_InvokeAsync.cs:
```diff
@@ -125,7 +189,15 @@ void WrappedCallback()
+                else
+                {
+                    completion.TrySetException(ex);
+                }
```
We can create a function to avoid repeating this pattern.

Shyam-Gupta reviewed (commented):
Just those two suggestions. Rest looks good to me.

KlausLoeffelmann committed:
Address review feedback

KlausLoeffelmann committed:
Fix non-static method.
